File: src/toolchain/host/platforms/ios/Sources/Z42VM/Z42VMValue.swift
import Foundation
import Z42VMC

/// A primitive value crossing the z42 host ABI. v0.1 supports the same
/// shapes as Tier 2 — null / i64 / f64 / bool. Strings / objects /
/// arrays land in H3 (see [`embedding.md §12 Deferred`]).
public enum Z42VMValue: Equatable {
    case null
    case i64(Int64)
    case f64(Double)
    case bool(Bool)
}

extension Z42VMValue {
    /// Convert to the raw `Z42Value` ABI struct.
    func toRaw() -> Z42Value {
        switch self {
        case .null:
            return Z42Value(tag: 0, reserved: 0, payload: 0)
        case let .i64(v):
            return Z42Value(tag: 1, reserved: 0, payload: UInt64(bitPattern: v))
        case let .f64(v):
            return Z42Value(tag: 2, reserved: 0, payload: v.bitPattern)
        case let .bool(v):
            return Z42Value(tag: 3, reserved: 0, payload: v ? 1 : 0)
        }
    }

    /// Convert from raw `Z42Value` returned by the runtime.
    /// Unsupported tags surface as `.null` (v0.1 limitation).
    static func fromRaw(_ raw: Z42Value) -> Z42VMValue {
        switch raw.tag {
        case 0: return .null
        case 1: return .i64(Int64(bitPattern: raw.payload))
        case 2: return .f64(Double(bitPattern: raw.payload))
        case 3: return .bool(raw.payload != 0)
        default: return .null
        }
    }
}

File: src/toolchain/host/platforms/android/z42vm/src/main/java/io/z42/vm/Z42VMValue.kt
package io.z42.vm

/**
 * Primitive value crossing the z42 host ABI. v0.1 supports the same
 * shapes as the cross-platform contract — null / i64 / f64 / bool.
 * Strings / objects / arrays land in a later spec.
 */
sealed class Z42VMValue {
    object Null : Z42VMValue()
    data class I64(val v: Long)    : Z42VMValue()
    data class F64(val v: Double)  : Z42VMValue()
    data class Bool(val v: Boolean): Z42VMValue()

    /** Tag dictionary mirroring `z42_abi.h Z42_VALUE_TAG_*`. */
    internal val tag: Int get() = when (this) {
        is Null -> 0
        is I64  -> 1
        is F64  -> 2
        is Bool -> 3
    }

    /**
     * 64-bit raw payload as the C ABI sees it. Float / bool are bit-
     * packed; null gets 0.
     */
    internal val payload: Long get() = when (this) {
        is Null -> 0L
        is I64  -> v
        is F64  -> java.lang.Double.doubleToRawLongBits(v)
        is Bool -> if (v) 1L else 0L
    }

    companion object {
        internal fun fromRaw(tag: Int, payload: Long): Z42VMValue = when (tag) {
            0 -> Null
            1 -> I64(payload)
            2 -> F64(java.lang.Double.longBitsToDouble(payload))
            3 -> Bool(payload != 0L)
            else -> Null // unsupported tags surface as null in v0.1
        }
    }
}
